[SPARK-23972][BUILD][SQL] Update Parquet to 1.10.0.

rdblue opened the pull request:
## What changes were proposed in this pull request?

This updates Parquet to 1.10.0 and updates the vectorized path for buffer management changes. Parquet 1.10.0 uses ByteBufferInputStream instead of byte arrays in encoders. This allows Parquet to break allocations into smaller chunks that are better for garbage collection.

## How was this patch tested?

Existing Parquet tests. Running in production at Netflix for about 3 months.

rdblue committed:
SPARK-23972: Update Parquet to 1.10.0.

This updates the vectorized path for changes in Parquet 1.10.0, which
uses ByteBufferInputStream instead of byte arrays in encoders. This
allows Parquet to break allocations into smaller chunks that are better
for garbage collection.

SparkQA commented:
**[Test build #89362 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89362/testReport)** for PR 21070 at commit [`4df17a6`](https://github.com/apache/spark/commit/4df17a6e9726cb22e499d479a9ab48f5db18a538).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/2327/
Test PASSed.

SparkQA commented:
**[Test build #89362 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89362/testReport)** for PR 21070 at commit [`4df17a6`](https://github.com/apache/spark/commit/4df17a6e9726cb22e499d479a9ab48f5db18a538).
 * This patch **fails build dependency tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89362/
Test FAILed.

gatorsmile commented:
Could you share the performance number?

rdblue commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedPlainValuesReader.java:
```diff
@@ -63,115 +58,139 @@ public final void readBooleans(int total, WritableColumnVector c, int rowId) {
+    int requiredBytes = total * 4;
+    ByteBuffer buffer = getBuffer(requiredBytes);
+
+    for (int i = 0; i < total; i += 1) {
```
It looks that way, but it actually replaces a similar loop: https://github.com/apache/spark/blob/master/sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/OnHeapColumnVector.java#L283-L291

The main problem is that ByteBufffer isn't supported in the column vectors. That seems beyond the scope of this PR.

henryr replied:
Couldn't it also be writing to an `OffHeapColumnVector`? https://github.com/apache/spark/blob/master/sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java#L199

If so, I think the copy is 1MB at a time: https://github.com/apache/spark/blob/master/common/unsafe/src/main/java/org/apache/spark/unsafe/Platform.java#L189

I agree that ByteBuffer shouldn't be supported in this PR. But there's an opportunity to use the bulk copy APIs which would benefit from any future optimization that happens. Plus even if the copy does eventually become a loop inside the column vector implementation, there's more chance of the JIT unrolling the loop since it's smaller.

kiszk replied:
I think that better integration between `ByteBuffer` and `ColumnVector` would be addressed in another PR.
Since tableCache also uses `ByteBuffer`, it would be good to discuss in another PR.
cc: @cloud-fan

henryr replied:
Agreed that fixing the `ByteBuffer` / `ColumnVector` interaction should be dealt with elsewhere. I'm just raising the possibility of _regressing_ the read path here because the copies are less efficient. Since it's going to be a while before 2.4.0, that might be ok if we commit to fixing it - but it superficially seems like a manageable change to the PR since the code to call the bulk APIs is already there.  What do you think?

rdblue replied:
The reason why I moved the loop out was that I didn't think that using the byte[] API would actually be better. Parquet doesn't guarantee that these byte buffers are on the heap and backed by byte arrays, so we would need to copy the bytes out of the buffer into an array only to copy them again in the column vector call. Two copies (and possibly allocation) seemed worse than moving the loop.

We could catch the case where the buffers are on-heap and make the bulk call. The drawback is that this will be temporary and will be removed when ColumnVector supports ByteBuffer. And, it only works/matters when Parquet uses on-heap buffers and Spark uses off-heap buffers. Is that worth the change to this PR? I can take a shot at it if you think it is. I'd rather update ColumnVector and then follow up though.

henryr replied:
Isn't that what `hasArray()` is for though? If the buffers are backed by a byte array, `hasArray()` returns true and accessing the byte array via `array()` should be 0 cost. (If `array()` actually copies any data, that would invalidate this line of reasoning but would also be unexpected).

So for example, here you'd have:
    
    public final void readIntegers(int total, WritableColumnVector c, int rowId) {
        int requiredBytes = total * 4;
        ByteBuffer buffer = getBuffer(requiredBytes);
        if (buffer.hasArray()) {
          c.putIntsLittleEndian(rowId, total, buffer.array(), 0);         
        } else {
            for (int i = 0; i < total; i += 1) {
                c.putInt(rowId + i, buffer.getInt());
            }
        }
    }

This seems to be the same pattern that's in `readBinary()`, below. Let me know if I'm missing something!

rdblue replied:
Yeah, exactly. We can detect that the buffer is backed by an array and use the other call. If we think this is worth it as a short-term fix, I'll update this PR.

rdblue replied:
Updated.

HyukjinKwon commented:
@rdblue, BTW mind fixing the title to `[SPARK-23972][...] ...`? It's actually written in the guide.

dongjoon-hyun commented on pom.xml:
```diff
@@ -129,7 +129,7 @@
     <hive.version.short>1.2.1</hive.version.short>
     <derby.version>10.12.1.1</derby.version>
-    <parquet.version>1.8.2</parquet.version>
+    <parquet.version>1.10.0</parquet.version>
```
@rdblue . To see the Jenkins result, could you resolve the dependency check failure with the following?
```
./dev/test-dependencies.sh --replace-manifest
```

dongjoon-hyun replied:
Unlike [last time](https://github.com/apache/spark/commit/26a4cba3ffaadf382ca14980378965704ccef9ab), it seems that this PR touches `commons-pool` dependency together? Can we avoid this?
```
-commons-pool-1.5.4.jar
+commons-pool-1.6.jar
```

rdblue replied:
I excluded the commons-pool dependency from parquet-hadoop to avoid this. I also tested the latest Parquet release with commons-pool 1.5.4 and everything passes. I don't think it actually requires 1.6.

rdblue committed:
SPARK-23972: Update dependencies for Parquet 1.10.0.

SparkQA commented:
**[Test build #89408 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89408/testReport)** for PR 21070 at commit [`8446b40`](https://github.com/apache/spark/commit/8446b40c243a012180b51260180542e03dd9a6c2).

SparkQA commented:
**[Test build #89408 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89408/testReport)** for PR 21070 at commit [`8446b40`](https://github.com/apache/spark/commit/8446b40c243a012180b51260180542e03dd9a6c2).
 * This patch **fails build dependency tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89408/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/2353/
Test PASSed.

SparkQA commented:
**[Test build #89409 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89409/testReport)** for PR 21070 at commit [`d724811`](https://github.com/apache/spark/commit/d724811fe97cf26c2140fe6fb07e862da1529c9f).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/2354/
Test PASSed.

SparkQA commented:
**[Test build #89409 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89409/testReport)** for PR 21070 at commit [`d724811`](https://github.com/apache/spark/commit/d724811fe97cf26c2140fe6fb07e862da1529c9f).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89409/
Test FAILed.

rdblue committed:
SPARK-23972: Fix method signatures for IOExceptions.

SparkQA commented:
**[Test build #89412 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89412/testReport)** for PR 21070 at commit [`839cf28`](https://github.com/apache/spark/commit/839cf282c4e8552620452171cec783aa7349ce5e).

rdblue committed:
SPAKR-23972: Bulk write to ColumnVector when Parquet data is on the heap.

SparkQA commented:
**[Test build #89415 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89415/testReport)** for PR 21070 at commit [`27a66d8`](https://github.com/apache/spark/commit/27a66d8114552e199389c944517d42719861b9de).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/2359/
Test PASSed.

henryr reviewed (commented):
Thanks for making this change! Looks pretty good to me, let's see what others have to say.

SparkQA commented:
**[Test build #89415 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89415/testReport)** for PR 21070 at commit [`27a66d8`](https://github.com/apache/spark/commit/27a66d8114552e199389c944517d42719861b9de).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89415/
Test FAILed.

scottcarey commented:
This PR should include changes to `ParquetOptions.scala` to expose the `LZ4`, `ZSTD` and `BROTLI` parquet compression codecs, or else spark users won't be able to leverage those parquet 1.10.0 features.

henryr commented:
@scottcarey I agree that's important. Perhaps it could be done as a follow-up PR?

rdblue commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedPlainValuesReader.java:
```diff
@@ -63,115 +59,159 @@ public final void readBooleans(int total, WritableColumnVector c, int rowId) {
-    }
-    offset += 8;
-    return v;
+    return getBuffer(8).getDouble();
```
Yeah, that's a good point. It isn't read this way in Parquet, but it would be useful for libraries that use this deep integration. Feel free to submit a PR on Parquet and I'll review it there.

rdblue committed:
SPARK-23972: Fix test failures and review comments.

rdblue committed:
SPARK-23972: Add lz4, brotli, and zstd compression for Parquet.

scottcarey commented:
I tested this with the addition of some changes to ParquetOptions.scala, but this alone does not allow for writing or reading zstd compressed parquet files, because it is using reflection to acquire hadoop classes for compression which are not in the supplied dependencies.

From what I can see, anyone that wants to use the new compression codecs are going to have to build their own custom version of spark.  And probably with modified versions of hadoop libraries as well, including changing how the native bindings are built.... because that would be easier than updating the whole thing to hadoop-common 3.0 where the required compressors exist.

Alternatively, spark+parquet should avoid the hadoop dependencies like the plague for compression / decompression.  They bring in a steaming heap of dependencies and possible library conflicts and users often have versions (or CDH versions) that don't exactly match.
In my mind, parquet should handle the compression itself, or with a light-weight dependency.
Perhaps it can use either the hadoop flavor, or if that is not found, another one, or even a user-supplied one so that it works stand-alone or from inside hadoop without issue.
Right now it is bound together with reflection and an awkward stack of brittle dependencies with no escape hatch.

Or am I missing something here, and it is possible to read/write with the new codecs if I configure it differently?

SparkQA commented:
**[Test build #89531 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89531/testReport)** for PR 21070 at commit [`5a78030`](https://github.com/apache/spark/commit/5a78030f2ba0b8df1098ccabad4afd30650d2336).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/2453/
Test PASSed.

rdblue committed:
SPARK-23972: Update InMemoryColumnarQuerySuite for new Parquet version.

The new Parquet version adds some metadata so files are slightly larger.
This updates the test case that reports the file size for the size of an
in-memory relation.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/2455/
Test PASSed.

scottcarey commented:
@rdblue 
The problem with zstd is that it is only in Hadoop 3.0, and dropping _that_ jar in breaks things as it is a major release.  Extracting out only the ZStandardCodec from that and recompiling to a 2.x release does not work either, because it depends on some low level hadoop native library management to load the native library (it does not appear to use  https://github.com/luben/zstd-jni).

The alternative is to write a custom ZStandardCodec implementation that uses luben:zstd-jni

Furthermore, if you add a `o.a.h.io.codecs.ZStandardCodec` class to a jar on the client side, it is still not found -- my guess is there is some classloader isolation between client code and spark itself and spark itself is what needs to find the class.  So one has to have it installed inside of the spark distribution.

I may take you up on fixing the compression codec dependency mess in a couple months.  The hardest part will be lining up the configuration options with what users already expect -- the raw codecs aren't that hard to do.

rdblue commented:
I backported the Hadoop zstd codec to 2.7.3 without much trouble. But either way, I think that's a separate concern. This is about getting Parquet updated, not about worrying whether users can easily add compression implementations to their classpath.

scottcarey commented:
> This is about getting Parquet updated, not about worrying whether users can easily add compression implementations to their classpath.

Yes, of course.

My hunch is that someone else will read the release notes that spark 2.3.0 supports zstandard, and parquet 1.10.0 supports zstandard, then realize it doesn't work in combination and end up here.  So I feel that this is the right place to discuss the state of these features until there is another more specific place to do so.

The discussion here has been useful to get closer to understanding what further tasks there may be.  If there are any follow-on issues, the discussion can move there.


I would love to be able to test this with my full use case, and give it a big thumbs up if it works.  Unfortunately my only motivation for this upgrade is access to ZStandard, and I'm not as excited to say 'works for me if I don't use new parquet codecs'.

SparkQA commented:
**[Test build #89531 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89531/testReport)** for PR 21070 at commit [`5a78030`](https://github.com/apache/spark/commit/5a78030f2ba0b8df1098ccabad4afd30650d2336).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89531/
Test FAILed.

SparkQA commented:
**[Test build #89533 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89533/testReport)** for PR 21070 at commit [`5fca3ce`](https://github.com/apache/spark/commit/5fca3ce1e199b0f3ee9d2b2178cf1f2418d4324f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89533/
Test FAILed.

rdblue committed:
SPARK-23972: Regenerate golden outputs for Parquet size changes.

SparkQA commented:
**[Test build #89600 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89600/testReport)** for PR 21070 at commit [`40a9cdd`](https://github.com/apache/spark/commit/40a9cdd969d3695a9dab5d6e18ed1b062be254bf).

SparkQA commented:
**[Test build #89600 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89600/testReport)** for PR 21070 at commit [`40a9cdd`](https://github.com/apache/spark/commit/40a9cdd969d3695a9dab5d6e18ed1b062be254bf).
 * This patch **fails from timeout after a configured wait of \`300m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89600/
Test FAILed.

rdblue commented:
Retest this please.

SparkQA commented:
**[Test build #89644 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89644/testReport)** for PR 21070 at commit [`40a9cdd`](https://github.com/apache/spark/commit/40a9cdd969d3695a9dab5d6e18ed1b062be254bf).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/2539/
Test PASSed.

SparkQA commented:
**[Test build #89644 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89644/testReport)** for PR 21070 at commit [`40a9cdd`](https://github.com/apache/spark/commit/40a9cdd969d3695a9dab5d6e18ed1b062be254bf).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89644/
Test PASSed.

cloud-fan commented:
Can we run a TPCDS and show that this upgrade doesn't cause performance regression in Spark? I can see that this new version doesn't have perf regression at parquet side, just want to be sure the Spark parquet integration is also OK.

rdblue commented:
@cloud-fan, is there a Jenkins job to run it?

cloud-fan commented:
No. But there are a bunch of open source projects to help you easily run TPCDS on Spark with parquet files, e.g. https://github.com/maropu/spark-tpcds-datagen

rdblue commented:
Okay, I don't have the time to set up and run benchmarks without a stronger case for this causing a regression (given the Parquet testing), but other people should feel free to pick this up.

henryr commented:
@cloud-fan since there’s probably quite some time before this lands in a release, what do you think about merging this now if it’s ready, and filing the perf jira as a blocker against 2.4? My guess is that Spark will want to move to 1.10 at some point no matter what, so doing it this way has the advantage of giving plenty of time for any other issues to shake out before the next release.

gatorsmile commented:
Based on the previous upgrade (e.g., https://github.com/apache/spark/pull/13280#issuecomment-223022415), we hit the performance regressions when we upgrade Parquet and we did the revert at the end. That might be the major reason why we have not upgraded to 1.9 yet. 

Jumping to 1.10.0 from 1.8.2 is big and risky especially when it was recently released. Even if we merge it now, we might still revert it if we hit ANY performance regression. That is why we prefer to do the performance testing before we do the merge. 

Not sure whether the community can help the performance tests? cc @maropu You did many performance tests in TPC-DS in the past release. Could you help this too? If possible, we also need to do the micro-benchmark like what @dongjoon-hyun did for ORC.

gatorsmile commented:
cc @liancheng @yhuai @rxin

henryr commented:
Ok, thanks for the context. I've worked on other projects where it's sometimes ok to take a small risk of a perf hit on trunk as long as the community is committed to addressing the issues before the next release. If the norms here are not to absorb any risk on trunk, that also seems reasonable.

maropu commented:
I'll check performance changes w/ this pr in days and put the result in https://issues.apache.org/jira/browse/SPARK-24070

gatorsmile commented:
Parquet is the default file format. Thus, the impact could be huge. We need to be more cautious. Since Parquet 1.10 was officially released this month. Thus, we might need more careful review for such a big version upgrade. 

@rdblue Could you summarize the major benefits brought by this upgrade from 1.8 to 1.10?

gatorsmile commented on sql/core/src/test/scala/org/apache/spark/sql/execution/columnar/InMemoryColumnarQuerySuite.scala:
```diff
@@ -503,7 +503,7 @@ class InMemoryColumnarQuerySuite extends QueryTest with SharedSQLContext {
           }.head
           // InMemoryRelation's stats is file size before the underlying RDD is materialized
-          assert(inMemoryRelation.computeStats().sizeInBytes === 740)
+          assert(inMemoryRelation.computeStats().sizeInBytes === 800)
```
Our optimizer uses the statistics to decide the plans (e.g., in join algorithm selection). Thus, the plans could be completely different if the file size increases by 8 percents. Could you give us more contexts? cc @rdblue

rdblue replied:
Parquet fixed a problem with value ordering in statistics, which required adding new metadata min and max fields. For older readers, Parquet also writes the old values when it makes sense to. This is a slight increase in overhead, which is more noticeable when files contain just a few records.

Don't be alarmed at the percentage difference here, it is just a small file. Parquet isn't increasing file sizes by 8%, that would be silly.

gatorsmile replied:
Then it looks fine, but the new metadata max/min fields are added in file metadata, column metadata and page header metadata? Any formula we can use to calculate the size increase?

rdblue replied:
This is data dependent so it is hard to estimate. We write the stats for older readers when the type uses a signed sort order, so it is limited to mostly primitive types and won't be written for byte arrays or utf8 data. That limits the size to 16 bytes + thrift overhead per page and you might have about 100 pages per row group. So 1.5k per 128MB, which is about 0.001%.

rdblue commented:
There are two main reasons to update. First, the problem behind SPARK-17213 is fixed, hence the new min and max fields. Second, this updates the internal byte array management, which is needed for page skipping in the next few versions.

gatorsmile commented:
Regarding [`SPARK-17213`](https://issues.apache.org/jira/browse/SPARK-17213), we already enable the pushdown after we upgrading to 1.8.2. Is my understanding right?

rdblue commented:
The fix for PARQUET-686 was to suppress min/max stats. It is safe to push filters, but those filters can't be used without the stats. 1.10.0 has the correct stats and can use those filters.

maropu commented:
@gatorsmile @rdblue I checked the numbers and I found no regression at least in TPC-DS: to check the actual numbers, see https://issues.apache.org/jira/browse/SPARK-24070.

rdblue commented:
Thank you @maropu!

What resources does the run require? Is it something we could create a Jenkins job to run?

maropu commented:
I just run `TPCDSQueryBenchmark` in an ec2 instance (m4.2xlarge). The job to check performance regression is too heavy, so we just check if `tpcds`, `tpch`, and `ssb` can be correctly compiled in `TPCDSQuerySuite`, `TPCHQuerySuite`, and `SSBQuerySuite`.

BTW, I sometime checks the regression in my repo by myself: https://github.com/maropu/spark-tpcds-datagen.

maropu commented:
ok, I'll work on it.

gatorsmile commented:
@iddoav Thank you for your feedbacks. It sounds like CDH5.X's spark has a bug, right? Our Apache Spark does not have such an issue based on my understanding. 

@rdblue Does that means we can see a better perf number after upgrading to 1.10.0, since it has the correct stats and can use these stats for string filter pushdown?

rdblue commented:
Yes, it is safe to use push-down for string columns in data written with this version.

gatorsmile commented:
@maropu Could you help verify the perf improvement in this string filter pushdown in the microbenchmark? You can also refer to another microbenchmark we did in ORC. https://github.com/apache/spark/blob/master/sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/OrcReadBenchmark.scala

If possible, please make a new microbenchmark suite that is extensible to all the data sources.

rdblue commented:
Is it necessary to block this commit on benchmarks? We know that it doesn't degrade performance from the Parquet benchmarks and TPC-DS run. What do you think needs to be done before this can move forward?

gatorsmile commented:
TPC-DS is just an end-to-end perf testing. To ensure no perf regressions, we need micro-benchmark tests. That is just a normal procedure. We  did the same in the ORC upgrade too.

mswit-databricks commented:
@rdblue Do you see any risk of additional overhead coming from the extra stats? For example, if the data contains very long strings, performing comparison on them to generate stats will be expensive and might not be worth it in certain scenarios.

rdblue commented:
@mswit-databricks, I wouldn't worry about that. We've limited the length of binary and string fields.

In the next version of Parquet, we're planning on releasing page indexes, which are lower and upper bounds instead of min and max values. That gives us more flexibility to shorten values and avoid the case that you're worried about.

maropu commented:
I'll put the benchmark results (the benchmark's currently running) today, just a sec.

maropu commented:
Simple scan benchmarks:
code: https://github.com/apache/spark/compare/master...maropu:DataSourceReadBenchmark
master: https://gist.github.com/maropu/a767d21ed1dd047ec2bdca92915dc5c5
this pr: https://gist.github.com/maropu/b9312488ec791caec9711ba64027e3e2

Pushdown benchmarks:
code: https://github.com/apache/spark/compare/master...maropu:UpdateParquetBenchmark
this pr: https://gist.github.com/maropu/e7cbfb8dc73a91806088e159f64e113f

Obviously, the pushdown for strings didn't work and I don't look into why now.
Can you check this? @rdblue

rdblue commented:
@maropu, are you sure about the INT and FLOAT columns? I think you might have that assessment backwards. Here's the INT results from the PR gist:

```
SQL Single INT Column Scan:              Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
SQL Parquet Vectorized                         149 /  162        105.5           9.5       1.0X
SQL Parquet MR                                1825 / 1836          8.6         116.1       0.1X
```

And here are the INT results from the master gist:

```
SQL Single INT Column Scan:              Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
SQL Parquet Vectorized                         250 /  292         63.0          15.9       1.0X
SQL Parquet MR                                3175 / 3202          5.0         201.8       0.1X
```

I think that shows that the PR result was significantly faster, not slower. (The other INT test was about the same.)

Here's the FLOAT column from the PR gist:

```
SQL Single FLOAT Column Scan:            Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
SQL Parquet Vectorized                         145 /  158        108.8           9.2       1.0X
SQL Parquet MR                                1840 / 1843          8.5         117.0       0.1X
```

And FLOAT from the master gist:

```
SQL Single FLOAT Column Scan:            Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
SQL Parquet Vectorized                         261 /  316         60.2          16.6       1.0X
SQL Parquet MR                                3267 / 3284          4.8         207.7       0.1X
```

Am I reading this incorrectly? I'm considering lower time values and higher rate values to be better.

rdblue commented:
@maropu, looking at the pushdown benchmark, it looks like ORC and Parquet either both benefit or both do not benefit from pushdown. In some cases ORC is much faster, which is due to the fact that ORC will skip reading pages, not just row groups. But, when ORC benefits from pushdown so does Parquet, for example the `Select 1 int row (value = 7864320)` case.

I think that you were expecting a string comparison case to have a significant benefit over non-pushdown. But I would only expect that if ORC had a similar benefit. That's because this is dependent on the clustering of values in the file so that Parquet can eliminate row groups. If ORC didn't have a benefit, then I would expect that the data just isn't clustered in a way that helps.

I'm not sure how you're generating data, but I'd recommend adding a sorted column case with enough data to create multiple row groups (or stripes for ORC). That would write data so that you can ignore some row groups and you should see a speed up.

Parquet also supports dictionary-based row group filtering. To test that, make sure you have a column that is entirely dictionary-encoded: pick a small set of values and randomly draw from that set. Then if you search for a value that isn't in that set you should see a speedup. Also make sure that you have `parquet.filter.dictionary.enabled=true` set in the Hadoop configuration so that Parquet uses dictionary filtering.

maropu commented:
@rdblue Aha, thanks for the explanation! 

> I think that you were expecting a string comparison case to have a significant benefit over 
> non->pushdown. But I would only expect that if ORC had a similar benefit.
> That's because this is dependent on the clustering of values in the file so that Parquet can
> eliminate row groups. If ORC didn't have a benefit, then I would expect that the data just
> isn't clustered in a way that helps.

The `string` case had the same test data set (monotonically-increasing ids) with the `int` case, but we didn't get the benefit of push-down only in the string case. Is the logic to eliminate row groups different between `int` cases and `string` cases in spite that we use the same dataset?
https://github.com/maropu/spark/blob/465aa420b1399aba7199aa2868ad6ae58d877d50/sql/core/src/test/scala/org/apache/spark/sql/FilterPushdownBenchmark.scala#L70

rdblue commented:
@maropu, I suspect that the problem is that comparison is different for strings: `"17297598712"` is less than `"5"` with string comparison.

maropu commented:
aha, I see and it might to be true. I'll check the benchmark again.

maropu commented:
@rdblue [I made the input test date sorted (clustered)](https://github.com/maropu/spark/blob/dfb6f3bad10058b28ff66a6d2b72225b8847296c/sql/core/src/test/scala/org/apache/spark/sql/FilterPushdownBenchmark.scala#L77) though, I didn't get a luck in the string case;
https://gist.github.com/maropu/b7ff6d528547de03a4a5cbabd120e78d
I'm still looking into why (I welcome any suggestion).

rdblue commented:
@maropu, I'd recommend looking at the Parquet files using [`parquet-cli`](http://search.maven.org/#search%7Cga%7C1%7Ca%3A%22parquet-cli%22) to see if you're getting reasonable min/max stats for your string columns. That will give you an idea of when predicate push-down will help.

maropu commented:
ok, give me more time to do so.

maropu commented:
ok, I got the reason why;
```

$ ./bin/spark-shell --master=local[1]

scala> import scala.util.Random
scala> :paste
def timer[R](f: => {}): Unit = {
  val count = 5
  val iters = (0 until count).map { i =>
    val t0 = System.nanoTime()
    f
    val t1 = System.nanoTime()
    val elapsed = t1 - t0 + 0.0
    println(s"#$i: ${elapsed / 1000000000.0}")
    elapsed
  }
  println("Avg. Elapsed Time: " + ((iters.sum / count) / 1000000000.0) + "s")
}
var dir = "/home/ec2-user/parquet-test-string"
val numRows = 1024 * 1024 * 15
val width = 6
val selectExpr = (1 to width).map(i => s"CAST(value AS STRING) c$i")
val valueCol = monotonically_increasing_id().cast("string")
val df = spark.range(numRows).map(_ => Random.nextLong).selectExpr(selectExpr: _*).withColumn("value", valueCol).sort("value")
df.write.mode("overwrite").parquet(dir)
spark.read.parquet(dir).createOrReplaceTempView("parquetTable")

scala> sql("SET spark.sql.parquet.filterPushdown=true")
scala> timer { sql("select * from parquetTable where value = '0'").collect }
#0: 1.041495043                                                                 
#1: 0.53017502
#2: 0.468505896
#3: 0.437655119
#4: 0.429151435
Avg. Elapsed Time: 0.5813965026s

scala> sql("select * from parquetTable where value = 0").explain
== Physical Plan ==
*(1) Project [c1#35, c2#36, c3#37, c4#38, c5#39, c6#40, value#41]
+- *(1) Filter (isnotnull(value#41) && (cast(value#41 as int) = 0))
   +- *(1) FileScan parquet [c1#35,c2#36,c3#37,c4#38,c5#39,c6#40,value#41] Batched: true, Format: Parquet, Location: InMemoryFileIndex[file:/home/ec2-user/parquet-test-string], PartitionFilters: [], PushedFilters: [IsNotNull(value)], ReadSchema: struct<c1:string,c2:string,c3:string,c4:string,c5:string,c6:string,value:string>

scala> timer { sql("select * from parquetTable where value = 0").collect }
#0: 10.656159769                                                                
#1: 10.583965537                                                                
#2: 10.486018192                                                                
#3: 10.475532898                                                                
#4: 10.494059857                                                                
Avg. Elapsed Time: 10.539147250600001s

scala> sql("select * from parquetTable where value = '0'").explain
== Physical Plan ==
*(1) Project [c1#35, c2#36, c3#37, c4#38, c5#39, c6#40, value#41]
+- *(1) Filter (isnotnull(value#41) && (value#41 = 0))
   +- *(1) FileScan parquet [c1#35,c2#36,c3#37,c4#38,c5#39,c6#40,value#41] Batched: true, Format: Parquet, Location: InMemoryFileIndex[file:/home/ec2-user/parquet-test-string], PartitionFilters: [], PushedFilters: [IsNotNull(value), EqualTo(value,0)], ReadSchema: struct<c1:string,c2:string,c3:string,c4:string,c5:string,c6:string,value:string>
```
We do push down the predicate `value = '0'` into the parquet filter though, we don't push down the predicate `value = 0`. So, this is a spark-side issue for push-down handling.

maropu commented:
I looked over the code and I fond this issue could be fix by;
https://github.com/apache/spark/compare/master...maropu:FixFilterPushdown
```
scala> sql("select * from parquetTable where value = 0").explain
== Physical Plan ==
*(1) Project [c1#35, c2#36, c3#37, c4#38, c5#39, c6#40, value#41]
+- *(1) Filter (isnotnull(value#41) && (cast(value#41 as int) = 0))
   +- *(1) FileScan parquet [c1#35,c2#36,c3#37,c4#38,c5#39,c6#40,value#41] Batched: true, Format: Parquet, Location: InMemoryFileIndex[file:/Users/maropu/Desktop/parquet-test], PartitionFilters: [], PushedFilters: [IsNotNull(value), EqualTo(value,0)], ReadSchema: struct<c1:string,c2:string,c3:string,c4:string,c5:string,c6:string,value:string>
```
But, I don't know this is a known issue and this fix is a right approach. cc: @HyukjinKwon

maropu commented:
ok, I checked string pushdown worked well;
https://gist.github.com/maropu/e7cbfb8dc73a91806088e159f64e113f

dongjoon-hyun commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -345,7 +345,7 @@ object SQLConf {
     .stringConf
     .transform(_.toLowerCase(Locale.ROOT))
-    .checkValues(Set("none", "uncompressed", "snappy", "gzip", "lzo"))
+    .checkValues(Set("none", "uncompressed", "snappy", "gzip", "lzo", "lz4", "brotli", "zstd"))
```
Could you update [sql-programming-guide.md](https://github.com/apache/spark/blame/master/docs/sql-programming-guide.md#L967) together?

rdblue replied:
Done.

HyukjinKwon commented:
Yup, https://github.com/apache/spark/pull/21070#issuecomment-386793202, just for clarification, the given attribute and literal is castable but they are not being as so, right?

I believe this is a known issue and there were several tries:

One approach was directly casting always and it was reverted (roughly 2 years ago?). Another approach was constant folding at optimizer level but it was rejected as it's too messy. Another approach was directly casting and comparing both values but it was also rejected since it sounded unsafe.

It was a long old story so probably worth double checking the history but I feel sure that I remember this story correctly.

The key point IIUC was that `translateFilter` should be super conservative and sounds we better need to check every possibility.

HyukjinKwon commented:
@liancheng should also remember all the story too since I talked with him at that time IIRC.

maropu commented:
@HyukjinKwon Thanks for the explanation! I feel it's ok to keep this as it it for now. btw, we don't need to document the story somewhere?

HyukjinKwon commented:
For documentation, I think it's fine. it's a trackable history in JIRA although it's a bit hard to search .. Also, there might be some diff we should check since I haven't checked the details of JIRAs and if it's still true.

gatorsmile commented:
> sql("select * from parquetTable where value = '0'")
> sql("select * from parquetTable where value = 0")

Semantically, they are different. The results should be also different. The plan shows whether these predicates are pushed down or not. 

Without the upgrade to 1.10.0, we can't get the benefit from the predicate pushdown between the above two queries. I also double checked it in my local environment. Now, it sounds like it is good reason to upgrade the version, since string predicates are pretty common. 

@rdblue @maropu Thank you for your great efforts and patience! This upgrade is definitely good to have. We should try to make it in the next release.

@maropu Could you also submit a separate PR for your micro benchmark? Thanks!

cc @cloud-fan @hvanhovell @mswit-databricks Do you have any comment about the implementation of this PR?

HyukjinKwon commented on dev/deps/spark-deps-hadoop-2.7:
```diff
@@ -163,13 +163,13 @@ orc-mapreduce-1.4.3-nohive.jar
+parquet-hadoop-1.10.0.jar
 parquet-hadoop-bundle-1.6.0.jar
-parquet-jackson-1.8.2.jar
+parquet-jackson-1.10.0.jar
```
(btw, don't forget to fix https://github.com/apache/spark/blob/ce7ba2e98e0a3b038e881c271b5905058c43155b/dev/deps/spark-deps-hadoop-3.1#L184-L190 too)

rdblue replied:
Done.

maropu commented:
@gatorsmile ok, I'll do later.

rdblue commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedPlainValuesReader.java:
```diff
@@ -63,115 +59,157 @@ public final void readBooleans(int total, WritableColumnVector c, int rowId) {
 
+  private ByteBuffer getBuffer(int length) {
+    try {
+      return in.slice(length).order(ByteOrder.LITTLE_ENDIAN);
```
No, `slice` doesn't copy. That's why we're using `ByteBuffer` now, to avoid copy operations.

Setting the byte order to `LITTLE_ENDIAN` is correct because it is for the buffer and Parquet buffers store values in little endian: https://github.com/apache/parquet-format/blob/master/Encodings.md.

rdblue commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedPlainValuesReader.java:
```diff
@@ -63,115 +59,157 @@ public final void readBooleans(int total, WritableColumnVector c, int rowId) {
+    int requiredBytes = total * 4;
+    ByteBuffer buffer = getBuffer(requiredBytes);
+
+    if (buffer.hasArray()) {
```
No, there is no guarantee that the buffer from Parquet is on the heap.

rdblue committed:
SPARK-23972: Update Hadoop 3.1 dependency profile.

rdblue committed:
SPARK-23972: Update sql-programming-guide.md.

SparkQA commented:
**[Test build #90327 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90327/testReport)** for PR 21070 at commit [`6c9d47b`](https://github.com/apache/spark/commit/6c9d47babd16b067923014d49b83bfd1afb33c9b).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/3007/
Test PASSed.

SparkQA commented:
**[Test build #90327 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90327/testReport)** for PR 21070 at commit [`6c9d47b`](https://github.com/apache/spark/commit/6c9d47babd16b067923014d49b83bfd1afb33c9b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/90327/
Test PASSed.

cloud-fan commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedPlainValuesReader.java:
```diff
@@ -63,115 +59,157 @@ public final void readBooleans(int total, WritableColumnVector c, int rowId) {
-    }
-    offset += 4;
-    return v;
+    return getBuffer(4).getInt();
```
will we have a performance problem here? i.e. creating `ByteBuffer` too frequently. cc @michal-databricks

cloud-fan replied:
I'd also assume `ByteBuffer.getInt` is slower than `Platform.getInt`.

cloud-fan replied:
maybe we can follow https://github.com/apache/spark/pull/21070/files#diff-ebb4762e900515a0cb6b0fc6327bb04cR569  ?

cloud-fan replied:
I'm just thinking https://github.com/apache/spark/pull/21070/files#diff-ebb4762e900515a0cb6b0fc6327bb04cR569 should be faster than `getBuffer(4).getInt()`, but either way should not be slower than the previous code.

This may not have a big performance impact and maybe a premature optimization, it's just my empiricism so it's up to you to decide to change it or not.

rdblue replied:
I think this should be better than `readIntLittleEndian` because that's doing length checks when it accesses the underlying byte array. Using `Platform.getInt` is a good idea, but I'd rather test that in a follow-up if that's alright, since we are confident that this as-is doesn't harm performance. We need to follow up with better unsafe support for byte buffers anyway.

cloud-fan commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedRleValuesReader.java:
```diff
@@ -619,32 +608,37 @@ private int ceil8(int value) {
+            // values are bit packed 8 at a time, so reading bitWidth will always work
+            ByteBuffer buffer = in.slice(bitWidth);
+            this.packer.unpack8Values(
+                buffer, buffer.arrayOffset() + buffer.position(), this.currentBuffer, valueIndex);
```
shall we assume the `ByteBuffer` may not be on-heap?

rdblue replied:
Good catch. Fixed to remove the call to `arrayOffset`. It should work with both on- and off-heap buffers now.

cloud-fan commented:
LGTM except 2 comments

HyukjinKwon commented:
(While I am here) seems fine to me otherwise.

rdblue committed:
Fix buffer unpacking offset in VectorizedRleValuesReader.

SparkQA commented:
**[Test build #90375 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90375/testReport)** for PR 21070 at commit [`95ecde0`](https://github.com/apache/spark/commit/95ecde09392ab4f2aab2f1165fb180c5466c26b4).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/3045/
Test PASSed.

SparkQA commented:
**[Test build #90375 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90375/testReport)** for PR 21070 at commit [`95ecde0`](https://github.com/apache/spark/commit/95ecde09392ab4f2aab2f1165fb180c5466c26b4).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/90375/
Test PASSed.

cloud-fan commented:
thanks, merging to master!
